ast/compile_test.go

```diff
@@ -1481,6 +1463,101 @@ func TestCompilerRewriteExprTerms(t *testing.T) {
 	}
 }
 
+func TestCompilerCheckDuplicateImports(t *testing.T) {
+	cases := []struct {
+		note           string
+		module         string
+		expectedErrors Errors
+		strict         bool
+	}{
+		{
+			note: "shadow",
+			module: `package test
+				import input.noconflict
+				import input.foo
+				import data.foo
+				import data.bar.foo
+			`,
+			expectedErrors: Errors{
+				&Error{
+					Location: NewLocation([]byte("import"), "", 4, 5),
+					Message:  "import must not shadow import input.foo",
+				},
+				&Error{
+					Location: NewLocation([]byte("import"), "", 5, 5),
+					Message:  "import must not shadow import input.foo",
+				},
+			},
+			strict: true,
+		}, {
+			note: "alias shadow",
+			module: `package test
+				import input.noconflict
+				import input.foo
+				import input.bar as foo
+			`,
+			expectedErrors: Errors{
+				&Error{
+					Location: NewLocation([]byte("import"), "", 4, 5),
+					Message:  "import must not shadow import input.foo",
+				},
+			},
+			strict: true,
+		}, {
+			note: "no strict",
+			module: `package test
+				import input.noconflict
+				import input.foo
+				import data.foo
+				import data.bar.foo
+				import input.bar as foo
+			`,
+			strict: false,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.note, func(t *testing.T) {
+			compiler := NewCompiler().WithStrict(tc.strict)
+			compiler.Modules = map[string]*Module{
+				"test": MustParseModule(tc.module),
+			}
+
+			compileStages(compiler, compiler.checkDuplicateImports)
+
+			if len(tc.expectedErrors) > 0 {
+				assertErrors(t, compiler.Errors, tc.expectedErrors, true)
+			} else {
+				assertNotFailed(t, compiler)
+			}
+		})
+	}
+}
+
+func assertErrors(t *testing.T, actual Errors, expected Errors, assertLocation bool) {
```

Comment:
I think we'll need a `t.Helper()` here. It'll make test failures report the line number of the call to `assertErrors`, not the line in the helper.